internal/engine/fswatch/watchmanager.go

```diff
@@ -237,65 +162,143 @@ func (w *WatchManager) OnChange(ctx context.Context, st store.RStore) {
-	// setup the watch first, to avoid a gap in coverage between setup and
-	// teardown. it's ok if we get a file event twice.
-	newWatches := make(map[model.TargetID]targetNotifyCancel)
-	for _, target := range setup {
-		logger := store.NewLogActionLogger(ctx, st.Dispatch)
-		ignore, err := w.createIgnoreMatcher(target)
-		if err != nil {
-			st.Dispatch(store.NewErrorAction(err))
-			continue
-		}
+	state := st.RLockState()
+	defer st.RUnlockState()
 
-		watcher, err := w.fsWatcherMaker(target.Dependencies(), ignore, logger)
-		if err != nil {
-			st.Dispatch(store.NewErrorAction(err))
-			continue
-		}
+	if !state.EngineMode.WatchesFiles() {
+		return
+	}
+
+	// TODO(milas): how can global ignores fit into the API model more cleanly?
+	newGlobalIgnores := globalIgnores(state)
+	watchable := FileWatchesForManifests(state.Manifests(), newGlobalIgnores)
+	specsToProcess := make(map[types.NamespacedName]filewatches.FileWatchSpec)
+	for _, w := range watchable {
+		specsToProcess[types.NamespacedName{Namespace: w.GetNamespace(), Name: w.GetName()}] = w.Spec
+	}
 
-		err = watcher.Start()
-		if err != nil {
-			st.Dispatch(store.NewErrorAction(err))
-			continue
+	if len(state.ConfigFiles) > 0 {
+		specsToProcess[types.NamespacedName{Name: ConfigsTargetID.String()}] = filewatches.FileWatchSpec{
+			WatchedPaths: state.ConfigFiles,
 		}
+	}
 
-		ctx, cancel := context.WithCancel(ctx)
-		go w.dispatchFileChangesLoop(ctx, target, watcher, st)
-		newWatches[target.ID()] = targetNotifyCancel{target, watcher, cancel}
+	for name, spec := range specsToProcess {
+		if tw, ok := w.targetWatches[name]; ok {
+			// compare against whatever the current watcher is using as a spec to determine if updates are needed
+			if equality.Semantic.DeepEqual(tw.spec, spec) {
+				continue
+			}
+			// since we're holding a reference to the old one, we just replace it in the map and clean up
+			// after the new watcher is successfully running to avoid missing any events
+			if err := w.addOrReplace(ctx, st, name, spec); err != nil {
+				logger.Get(ctx).Debugf("Error updating watch for %q: %v", name.String(), err)
+				continue
+			}
+			w.cleanupWatch(ctx, tw)
+
+			fw := state.FileWatches[name]
+			if fw == nil {
+				// HACK: since store actions are handled asynchronously, if a spec immediately changes
+				// after creation, we can end up here before it's visible in the store; for now, creating
+				// a new one is fine, as this should really only happen during startup as various configs
+				// are loaded in and layered in rapid succession
+				fw = &filewatches.FileWatch{
+					ObjectMeta: metav1.ObjectMeta{Namespace: name.Namespace, Name: name.Name},
+					Spec:       *spec.DeepCopy(),
+				}
+			} else {
+				fw = fw.DeepCopy()
+				spec.DeepCopyInto(&fw.Spec)
+			}
+
+			// we always have the latest status, so just unconditionally overwrite whatever is in the object
+			tw.status.DeepCopyInto(&fw.Status)
+
+			st.Dispatch(NewFileWatchUpdateAction(fw))
+		} else {
```

Comment:
i think that if you make create/update idempotent, you might be able to merge these two branches

Reply:
Ah, good call - I had gone that route in the other PR but for some reason didn't do the same here. Just pushed some changes that make this dramatically more straightforward 😛